File: gpu/examples/hole.cpp
#include <geometries/psMakeHole.hpp>
#include <models/psgSF6O2Etching.hpp>
#include <process/psProcess.hpp>
#include <psUtil.hpp>

using namespace viennaps;

int main(int argc, char **argv) {
  using NumericType = double;
  constexpr int D = 3;

  Logger::setLogLevel(LogLevel::TIMING);
  omp_set_num_threads(16);

  // Parse the parameters
  util::Parameters params;
  if (argc > 1) {
    params.readConfigFile(argv[1]);
  } else {
    params.readConfigFile("config.txt");

    // std::cout << "Usage: " << argv[0] << " <config file>" << std::endl;
    // return 1;
  }

  auto context = DeviceContext::createContext();

  // set parameter units
  units::Length::setUnit(params.get<std::string>("lengthUnit"));
  units::Time::setUnit(params.get<std::string>("timeUnit"));

  // geometry setup
  auto geometry = Domain<NumericType, D>::New(
      params.get("gridDelta"), params.get("xExtent"), params.get("yExtent"));
  geometry->enableMetaData();
  MakeHole<NumericType, D>(geometry, params.get("holeRadius"),
                           0.0, // holeDepth
                           0.0, // holeTaperAngle
                           params.get("maskHeight"), params.get("taperAngle"),
                           HoleShape::HALF)
      .apply();

  // use pre-defined model SF6O2 etching model
  auto modelParams = SF6O2Etching<NumericType, D>::defaultParameters();
  modelParams.ionFlux = params.get("ionFlux");
  modelParams.etchantFlux = params.get("etchantFlux");
  modelParams.passivationFlux = params.get("oxygenFlux");
  modelParams.Ions.meanEnergy = params.get("meanEnergy");
  modelParams.Ions.sigmaEnergy = params.get("sigmaEnergy");
  modelParams.Ions.exponent = params.get("ionExponent");
  modelParams.Passivation.A_ie = params.get("A_O");
  modelParams.Substrate.A_ie = params.get("A_Si");
  modelParams.etchStopDepth = params.get("etchStopDepth");
  auto model =
      SmartPointer<gpu::SF6O2Etching<NumericType, D>>::New(modelParams);

  RayTracingParameters<D> rayTracingParams;
  rayTracingParams.raysPerPoint = params.get<unsigned>("raysPerPoint");
  rayTracingParams.smoothingNeighbors = 2;

  CoverageParameters coverageParams;
  coverageParams.maxIterations = 20;
  coverageParams.coverageDeltaThreshold = 1e-4;

  AdvectionParameters advParams;
  advParams.integrationScheme = util::convertIntegrationScheme(
      params.get<std::string>("integrationScheme"));

  // process setup
  Process<NumericType, D> process(geometry, model);
  process.setProcessDuration(params.get("processTime"));
  process.setRayTracingParameters(rayTracingParams);
  process.setCoverageParameters(coverageParams);
  process.setAdvectionParameters(advParams);
  process.setFluxEngineType(FluxEngineType::GPU_TRIANGLE);
  process.setDeviceId(0);

  // print initial surface
  geometry->saveSurfaceMesh("initial.vtp");

  // run the process
  process.apply();

  // print final surface
  geometry->saveSurfaceMesh(params.get<std::string>("outputFile"));
}

File: gpu/examples/hole.py
from argparse import ArgumentParser
import viennaps.d3 as psd
import viennaps as ps

# parse config file name and simulation dimension
parser = ArgumentParser(prog="hole", description="Run a hole etching process.")
parser.add_argument("filename")
args = parser.parse_args()

params = ps.ReadConfigFile(args.filename)

# print intermediate output surfaces during the process
ps.Logger.setLogLevel(ps.LogLevel.INFO)
ps.setNumThreads(16)

ps.Length.setUnit(params["lengthUnit"])
ps.Time.setUnit(params["timeUnit"])

ps.gpu.Context.createContext()

# geometry setup, all units in um
geometry = psd.Domain(
    gridDelta=params["gridDelta"],
    xExtent=params["xExtent"],
    yExtent=params["yExtent"],
)
psd.MakeHole(
    domain=geometry,
    holeRadius=params["holeRadius"],
    holeDepth=0.0,
    maskHeight=params["maskHeight"],
    maskTaperAngle=params["taperAngle"],
    holeShape=ps.HoleShape.HALF,
).apply()

# use pre-defined model SF6O2 etching model
modelParams = psd.SF6O2Etching.defaultParameters()
modelParams.ionFlux = params["ionFlux"]
modelParams.etchantFlux = params["etchantFlux"]
modelParams.passivationFlux = params["oxygenFlux"]
modelParams.Ions.meanEnergy = params["meanEnergy"]
modelParams.Ions.sigmaEnergy = params["sigmaEnergy"]
modelParams.Ions.exponent = params["ionExponent"]
modelParams.Passivation.A_ie = params["A_O"]
modelParams.Substrate.A_ie = params["A_Si"]
modelParams.etchStopDepth = params["etchStopDepth"]
model = psd.SF6O2Etching(modelParams)

rayTracingParams = ps.RayTracingParameters()
rayTracingParams.smoothingNeighbors = 2
rayTracingParams.raysPerPoint = int(params["raysPerPoint"])

advectionParams = ps.AdvectionParameters()
advectionParams.integrationScheme = ps.util.convertIntegrationScheme(
    params["integrationScheme"]
)

# process setup
process = psd.Process()
process.setDomain(geometry)
process.setProcessModel(model)
process.setProcessDuration(params["processTime"])  # seconds
process.setRayTracingParameters(rayTracingParams)
process.setAdvectionParameters(advectionParams)
process.setFluxEngineType(ps.FluxEngineType.GPU_TRIANGLE)

# print initial surface
geometry.saveSurfaceMesh(filename="initial.vtp", addMaterialIds=True)

# run the process
process.apply()

# print final surface
geometry.saveSurfaceMesh(filename="final.vtp", addMaterialIds=True)
